pkg/reconciler/buildrun/resources/taskrun_test.go

```diff
@@ -181,6 +186,80 @@ var _ = Describe("GenerateTaskrun", func() {
+				Expect(err).To(BeNil())
+
+				buildRunWithEnvs, err = ctl.LoadBuildRunFromBytes([]byte(test.MinimalBuildahBuildRunWithEnvVars))
+				Expect(err).To(BeNil())
+
+				buildStrategy, err = ctl.LoadBuildStrategyFromBytes([]byte(test.MinimalBuildahBuildStrategy))
+				Expect(err).To(BeNil())
+				buildStrategy.Spec.BuildSteps[0].ImagePullPolicy = "Always"
+
+				buildStrategyWithEnvs, err = ctl.LoadBuildStrategyFromBytes([]byte(test.MinimalBuildahBuildStrategyWithEnvs))
+				Expect(err).To(BeNil())
+				buildStrategy.Spec.BuildSteps[0].ImagePullPolicy = "Always"
+
+				expectedCommandOrArg = []string{
+					"bud", "--tag=$(params.shp-output-image)", fmt.Sprintf("--file=$(inputs.params.%s)", "DOCKERFILE"), "$(params.shp-source-context)",
+				}
+			})
+
+			It("should contain env vars specified in Build in every step", func() {
+				got, err = resources.GenerateTaskSpec(config.NewDefaultConfig(), buildWithEnvs, buildRun, buildStrategy.Spec.BuildSteps, []buildv1alpha1.Parameter{})
+				Expect(err).To(BeNil())
+
+				combinedEnvs, err := env.MergeEnvVars(buildRun.Spec.Env, buildWithEnvs.Spec.Env, true)
+				Expect(err).NotTo(HaveOccurred())
+
+				for _, step := range got.Steps {
+					Expect(len(step.Env)).To(Equal(len(combinedEnvs)))
+					Expect(reflect.DeepEqual(combinedEnvs, step.Env)).To(BeTrue())
+				}
+			})
+
+			It("should contain env vars specified in BuildRun in every step", func() {
+				got, err = resources.GenerateTaskSpec(config.NewDefaultConfig(), build, buildRunWithEnvs, buildStrategy.Spec.BuildSteps, []buildv1alpha1.Parameter{})
+				Expect(err).To(BeNil())
+
+				combinedEnvs, err := env.MergeEnvVars(buildRunWithEnvs.Spec.Env, build.Spec.Env, true)
+				Expect(err).NotTo(HaveOccurred())
+
+				for _, step := range got.Steps {
+					Expect(len(step.Env)).To(Equal(len(combinedEnvs)))
+					Expect(reflect.DeepEqual(combinedEnvs, step.Env)).To(BeTrue())
+				}
+			})
+
+			It("should override Build env vars with BuildRun env vars in every step", func() {
+				got, err = resources.GenerateTaskSpec(config.NewDefaultConfig(), buildWithEnvs, buildRunWithEnvs, buildStrategy.Spec.BuildSteps, []buildv1alpha1.Parameter{})
+				Expect(err).To(BeNil())
+
+				combinedEnvs, err := env.MergeEnvVars(buildRunWithEnvs.Spec.Env, buildWithEnvs.Spec.Env, true)
+				Expect(err).NotTo(HaveOccurred())
+
+				for _, step := range got.Steps {
+					Expect(len(step.Env)).To(Equal(len(combinedEnvs)))
+					Expect(reflect.DeepEqual(combinedEnvs, step.Env)).To(BeTrue())
+
+				}
+			})
+
+			It("should fail attempting to override an env var in a BuildStrategy", func() {
+				got, err = resources.GenerateTaskSpec(config.NewDefaultConfig(), buildWithEnvs, buildRunWithEnvs, buildStrategyWithEnvs.Spec.BuildSteps, []buildv1alpha1.Parameter{})
+				Expect(err).NotTo(BeNil())
+				Expect(err.Error()).To(Equal("error(s) occurred in step \"buildah-bud\" of BuildStrategy \"buildah\": environment variable \"MY_VAR_1\" already exists"))
+			})
```

Comment:
again lots of good permutations here @coreydaley @adambkaplan but from what I can tell, we still do not have all the source options vs. the simple key/value pair for corev1.EnvVar

Reply:
updated in next push